new/target_mode.md

```diff
@@ -0,0 +1,53 @@
+---
+RFC: unassigned
+Title: Chef Target Mode
+Author: Thom May <thom@chef.io>
+Status: Draft
+Type: Standards Track
+---
+
+# Chef Target Mode
+
+When deploying applications or systems, we frequently need to configure
+appliances where Chef cannot run natively. To allow us to do so, we need
+to run chef on a dedicated management node and interact with the
+appliance remotely, via an API.
+
+## Motivation
+
+    As a site reliability engineer,
+    I want to configure routers, switches and load balancers,
+    so that my application functions correctly.
+
+## Specification
+
+The concept of a node in chef is a 1:1 mapping with a client, enabling a
+single node to have several identities if needed. We can leverage this
+facility to create a target mode, allowing a management node to run chef
+for many appliances.
+Target mode is enabled by several new configuration options:
+ * `target`: The details for connectivity to the target
+   system. The result of this configuration option should be a
+   class instance that cookbooks can send messages to. This
+   enables target mode, disabling ohai entirely.
+ * `platform`: The platform that we're connecting to. Since we don't
+   have ohai, this can be used to make decisions in a cookbook.
+
+Since target mode does not affect the local node, we can run many target
+mode chef client runs in parallel. Thus, a chef client run should only
+block other chef client runs that affect the same target, rather than
+being system wide.
```

Comment:
This should probably just call out directly that we need to modify the run lock around the chef-client to have per-target filenames so that they do not conflict with each other.  That is an implementation detail, but its the kind of obvious work that we'd need to do in order to make this more pleasant to work with.

Reply:
At Facebook we do this with a wrapper that takes a target (in our case phone IDs), and uses separate lockfiles, and uses that to pass in a separate set of attributes.json files to Chef to bootstrap itself.

Reply:
I don't really think the lock file is the issue here. That's also really easy to work-around by changing the `lockfile` in `client.rb`.

I think it's pretty difficult to enforce that one chef run will only touch the remote system and not the local system.
In the case of android phones, it makes a lot of sense to control the installed version of the android toolchain with the code that interacts with the phones, since they usually need to change in lock-step. But once you start managing files on the local system, you have to have some way to deconflict the operating-system level chef run and the chef run managing the phones.